ASFAchievementManager/AchievementHandler.cs

```diff
@@ -46,34 +44,27 @@ internal sealed class GetAchievementsCallback : CallbackMsg {
 				}
 
 				JobID = jobID;
-				Success = (EResult) msg.eresult == EResult.OK;
+				Success = (EResult) eresult(msg) == EResult.OK;
 				Response = msg;
 
 				if (!Success) {
-					ASF.ArchiLogger.LogGenericError(string.Format(Strings.ErrorFailingRequest, "GetAchievements"));
+					LogFailure(error);
 				}
 			}
-		}
-
-		internal sealed class SetAchievementsCallback : CallbackMsg {
-			internal readonly bool Success;
 
-			internal SetAchievementsCallback(JobID jobID, CMsgClientStoreUserStatsResponse msg) {
-				if (jobID == null) {
-					throw new ArgumentNullException(nameof(jobID));
-				}
-
-				if (msg == null) {
-					throw new ArgumentNullException(nameof(msg));
-				}
+			protected void LogFailure(string error) {
+				ASF.ArchiLogger.LogGenericError(string.Format(Strings.ErrorFailingRequest, error));
+			}
+		}
 
-				JobID = jobID;
-				Success = (EResult) msg.eresult == EResult.OK;
+		internal sealed class GetAchievementsCallback : AchievementsCallBack<CMsgClientGetUserStatsResponse> {
+			internal GetAchievementsCallback(JobID jobID, CMsgClientGetUserStatsResponse msg, Func<CMsgClientGetUserStatsResponse, int> eresult)
```

Comment:
remove the `eresult` parameter here and put the value you used above into the base call. This should be

 `internal GetAchievementsCallback(JobID jobID, CMsgClientGetUserStatsResponse msg) : base(jobID, msg, msg => msg.eresult, "SetArchievements")`)

Reply:
That is cleaner - thanks.

Reply:
It also allows the compiler to inline this call into the class instead of having to store it for every instance and thus is a performance improvement ;)